imu_integrator/plugin.cpp

```diff
@@ -0,0 +1,424 @@
+class imu_integrator : public threadloop {
+public:
+	imu_integrator(std::string name_, phonebook* pb_)
+		: threadloop{name_, pb_}
+		, sb{pb->lookup_impl<switchboard>()}
+		, _m_imu_cam{sb->subscribe_latest<imu_cam_type>("imu_cam")}
+		, _m_in{sb->subscribe_latest<imu_integrator_seq>("imu_integrator_seq")}
+		, _m_imu_integrator_input{sb->subscribe_latest<imu_integrator_input>("imu_integrator_input")}
+		, _m_imu_raw{sb->publish<imu_raw_type>("imu_raw")}
+		, _seq_expect(1)
+	{}
+
+	virtual skip_option _p_should_skip() override {
+		auto in = _m_in->get_latest_ro();
+		if (!in || in->seq == _seq_expect-1) {
+			// No new data, sleep to keep CPU utilization low
+			std::this_thread::sleep_for(std::chrono::milliseconds{1});
+			return skip_option::skip_and_yield;
+		} else {
+			if (in->seq != _seq_expect) {
+				_stat_missed = in->seq - _seq_expect;
+			} else {
+				_stat_missed = 0;
+			}
+			_stat_processed++;
+			_seq_expect = in->seq+1;
+			return skip_option::run;
+		}
+	}
+
+	void _p_one_iteration() override {
+		const imu_cam_type *datum = _m_imu_cam->get_latest_ro();
+		double timestamp_in_seconds = (double(datum->dataset_time) / NANO_SEC);
+
+		imu_type data;
+        data.timestamp = timestamp_in_seconds;
+        data.wm = (datum->angular_v).cast<double>();
+        data.am = (datum->linear_a).cast<double>();
+		_imu_vec.emplace_back(data);
+
+		clean_imu_vec(timestamp_in_seconds);
+        propagate_imu_values(timestamp_in_seconds, datum->time);
+	}
+
+private:
+	const std::shared_ptr<switchboard> sb;
+
+	// IMU Data, Sequence Flag, and State Vars Needed
+	std::unique_ptr<reader_latest<imu_cam_type>> _m_imu_cam;
+	std::unique_ptr<reader_latest<imu_integrator_seq>> _m_in;
+	std::unique_ptr<reader_latest<imu_integrator_input>> _m_imu_integrator_input;
+
+	// IMU Biases
+	std::unique_ptr<writer<imu_raw_type>> _m_imu_raw;
+	std::vector<imu_type> _imu_vec;
+	double last_imu_offset;
+	bool has_last_offset = false;
+	int64_t _seq_expect, _stat_processed, _stat_missed;
+
+	[[maybe_unused]] int counter = 0;
+	[[maybe_unused]] int cam_count = 0;
+	[[maybe_unused]] int total_imu = 0;
+	[[maybe_unused]] double last_cam_time = 0;
+
+	// Open_VINS cleans IMU values older than 20 seconds, we clean values older than 5 seconds
+	void clean_imu_vec(double timestamp) {
+		auto it0 = _imu_vec.begin();
+        while (it0 != _imu_vec.end()) {
+            if (timestamp-(*it0).timestamp > IMU_SAMPLE_LIFETIME) {
+                it0 = _imu_vec.erase(it0);
+            } else {
+                it0++;
+            }
+         }
+	}
+
+	// Timestamp we are propagating the biases to (new IMU reading time)
+	void propagate_imu_values(double timestamp, time_type real_time) {
+		const imu_integrator_input *input_values = _m_imu_integrator_input->get_latest_ro();
+		if (input_values == NULL) {
+			return;
+		}
+
+		if (!has_last_offset) {
+			last_imu_offset = input_values->t_offset;
+			has_last_offset = true;
+		}
+
+		Eigen::Matrix<double, 16, 1> imu_value;
```

Comment:
Is there a better name for this variable? This one is too misleading in this context because IMU values could also be the actual readings from the IMU.